### PR TITLE
ref(server): Change internal flagging for metric extraction

### PR DESCRIPTION
**Pull request by jan-auer**
Restructures the signatures of the metric extraction functions so that there are
distinct top-level functions that can be called and tested independently of each
other. These functions are:

 - generic metric extraction
 - transaction metric extraction
 - conditional tagging for transaction metrics
 - span metric extraction

All of these functions are called from a single place `extract_metrics` in the
envelope processor. The transaction metrics flag is checked a single time.

#skip-changelog
Closes https://github.com/getsentry/team-ingest/issues/163

### CODE REVIEW

File: relay-server/src/actors/processor.rs [hunk 1]
```diff
@@ -2064,83 +2057,74 @@ impl EnvelopeProcessorService {
...
-        }
+    /// Extract metrics from all envelope items.
+    ///
+    /// Caveats:
+    ///  - This functionality is incomplete. At this point, extraction is implemented only for
+    ///    transaction events.
+    fn extract_metrics(&self, state: &mut ProcessEnvelopeState) -> Result<(), ProcessingError> {
+        // TODO: Make span metrics extraction immutable
+        if let Some(event) = state.event.value_mut() {
+            if state.transaction_metrics_extracted {
```
Comment (jjbayer) [inline relay-server/src/actors/processor.rs]: Should we go ahead and rename this field to `event_metrics_extracted`?

File: relay-server/src/actors/processor.rs [hunk 2]
```diff
@@ -2064,83 +2057,74 @@ impl EnvelopeProcessorService {
...
-    ///  - This functionality is incomplete. At this point, extraction is implemented only for
-    ///    transaction events.
-    ///  - This function must run BEFORE `extract_transaction_metrics` to avoid double-extraction.
-    fn extract_metrics(&self, state: &mut ProcessEnvelopeState) -> Result<(), ProcessingError> {
-        let config = match state.project_state.config.metric_extraction {
-            ErrorBoundary::Ok(ref config) if config.is_enabled() => config,
-            _ => return Ok(()),
-        };
+                    state.extracted_metrics.extend(extracted);
+                    state.transaction_metrics_extracted |= true;
```
Comment (jjbayer) [inline relay-server/src/actors/processor.rs]: I think this makes sense in preparation of replacing transaction metrics extraction (partially) into `extract_event_metrics`. But we have to be careful not to enable `extract_event_metrics` without `extract_transaction_metrics` (e.g. in self-hosted).

**Review by jjbayer — APPROVED** [review]
LGTM. Integration test fails because we now set the flag even if only `extract_event_metrics` runs.

File: relay-server/src/actors/processor.rs (hunk 1 above)
Comment (jan-auer) [reply relay-server/src/actors/processor.rs]: Yeah, let's go ahead with this.

File: relay-server/src/actors/processor.rs (hunk 2 above)
Comment (jan-auer) [reply relay-server/src/actors/processor.rs]: Makes sense, this is a great callout. Currently, we have this coupling in Sentry [here](https://github.com/getsentry/sentry/blob/27ba845bb6b9e4f5169d0513966ab479435e2495/src/sentry/relay/config/__init__.py#L376). Metric extraction config is only generated if transaction metrics are generated.

File: relay-server/src/metrics_extraction/transactions/mod.rs [hunk 3]
```diff
@@ -231,194 +229,177 @@ fn extract_universal_tags(event: &Event, config: &TransactionMetricsConfig) -> C
...
-            has_profile,
-        },
-        project_metrics,
-        sampling_metrics,
-    )?;
-
-    if extract_spans_metrics {
-        extract_span_metrics(aggregator_config, event, project_metrics)?;
-    }
+/// TODO(ja): Doc
```
Comment (olksdr) [inline relay-server/src/metrics_extraction/transactions/mod.rs:232]: Are you going to address this TODO in the followup ?

File: relay-server/src/actors/processor.rs [hunk 4]
```diff
@@ -220,16 +221,8 @@ fn has_unprintable_fields(event: &Annotated<Event>) -> bool {
...
-struct ExtractedMetrics {
-    /// Metrics associated with the project that the transaction belongs to.
-    project_metrics: Vec<Metric>,
-    /// Metrics associated with the sampling project (a.k.a. root or head project)
-    /// which started the trace. See [`ProcessEnvelopeState::sampling_project_state`].
-    sampling_metrics: Vec<Metric>,
-}
-
 impl ExtractedMetrics {
+    // TODO(ja): Move
```
Comment (olksdr) [inline relay-server/src/actors/processor.rs:225]: Is this TODO for the followup?

**Review by olksdr — COMMENTED** [review]
There are still few TODOs left, like adding docs or moving things around. 
Are you going to address those in the followups?

**Pull request merged**